packages/react-query/src/shared/hooks/createHooksInternal.tsx

```diff
@@ -273,12 +273,37 @@ export function createRootHooks<
       queryClient.getMutationDefaults(mutationKey),
     );
 
+    // the result of `data` should be an array if we are using an async iterable
+    const [dataOverride, setDataOverride] = React.useState<unknown[] | null>(
+      null,
+    );
     const hook = __useMutation(
       {
         ...opts,
         mutationKey: mutationKey,
-        mutationFn: (input) => {
-          return client.mutation(...getClientArgs([path, { input }], opts));
+        mutationFn: async (input) => {
+          setDataOverride(null);
+          const result = await client.mutation(
+            ...getClientArgs([path, { input }], opts),
+          );
+
+          if (!isAsyncIterable(result)) {
+            return result;
+          }
+
+          // `mutateAsync()` returns an async iterable
+          async function* iterable() {
+            setDataOverride([]);
+            for await (const value of result as AsyncIterable<unknown>) {
+              setDataOverride((prev) => {
+                const next = [...(prev ?? []), value];
+                return next;
+              });
+              yield value;
+            }
+          }
+
+          return iterable();
```

Comment:
cc @TkDodo  👀 

is there a way for me to make the mutation stay as _pending_ but still have `mutateAsync` return the result?

tl;dr

```ts
const mutation = trpc.iterable.useMutation()

mutation.data // <--- should be an array that gets updated

// [...]
await mutation.mutateAsync(); // <--- should return an `AsyncIterable`
```

I'm not sure how to access the `MutationObserver` within the function here

Reply:
I don't think so. pending state is defined as "we have no data".

> I'm not sure how to access the MutationObserver within the function here

It's not exposed. You would probably have to roll your own `useMutation`. If you look at the code, it doesn't do that much: Basically create an observer, subscribe to it in `useSyncExternalStore` and return the result ...